src/coreclr/tools/Common/JitInterface/RISCV64PassStructInRegister.cs

```diff
@@ -1,210 +1,125 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 using System;
+using System.Collections.Generic;
 using System.Diagnostics;
 using ILCompiler;
 using Internal.TypeSystem;
+using static Internal.JitInterface.StructFloatFieldInfoFlags;
 
 namespace Internal.JitInterface
 {
-
     internal static class RISCV64PassStructInRegister
     {
-        public static uint GetRISCV64PassStructInRegisterFlags(TypeDesc typeDesc)
+        private const int
+            ENREGISTERED_PARAMTYPE_MAXSIZE = 16,
+            TARGET_POINTER_SIZE = 8;
+
+        private static bool HandleInlineArray(int elementTypeIndex, int nElements, StructFloatFieldInfoFlags[] types, ref int typeIndex)
         {
-            FieldDesc firstField = null;
-            uint floatFieldFlags = (uint)StructFloatFieldInfoFlags.STRUCT_NO_FLOAT_FIELD;
-            int numIntroducedFields = 0;
-            foreach (FieldDesc field in typeDesc.GetFields())
+            int nFlattenedFieldsPerElement = typeIndex - elementTypeIndex;
+            if (nFlattenedFieldsPerElement == 0)
+                return true;
+
+            Debug.Assert(nFlattenedFieldsPerElement == 1 || nFlattenedFieldsPerElement == 2);
+
+            if (nElements > 2)
+                return false;
+
+            if (nElements == 2)
             {
-                if (!field.IsStatic)
-                {
-                    firstField ??= field;
-                    numIntroducedFields++;
-                }
+                if (typeIndex + nFlattenedFieldsPerElement > 2)
+                    return false;
+
+                Debug.Assert(elementTypeIndex == 0);
+                Debug.Assert(typeIndex == 1);
+                types[typeIndex++] = types[elementTypeIndex]; // duplicate the array element type
             }
+            return true;
+        }
 
-            if ((numIntroducedFields == 0) || (numIntroducedFields > 2) || (typeDesc.GetElementSize().AsInt > 16))
+        private static bool FlattenFieldTypes(TypeDesc td, StructFloatFieldInfoFlags[] types, ref int typeIndex)
```

Comment:
Sure.

There's still more opportunities, like use a single ref to `struct { StructFloatFieldInfoFlags[2] types, int index; }` (or even pass & return by value, it's so small). Since I'll be revisiting this in another PR to handle corner-cases with empty fields, I'll probably do sth like this in both C# and C++ while at it.